Field operations

willsimmMOA opened the pull request:
added:

WorkRecordResource
OperationRecordResource
OperationEnvironmentType*
OperationProductType*
ComponentProductType
ProductResource*
ProductAnalysisType
ProductIngredientType
RegistrationType
WithdrawalPeriodType
....and associated enumerations

*Still to work up QuantityMeasurementType and ClimateMeasurement after some clarification

ab#15830

willsimmMOA committed:
Create FeatureReferenceType.json

willsimmMOA committed:
Create ObservationType.json

willsimmMOA committed:
Create workItemPriority.json

willsimmMOA committed:
Create workItemStatus.json

willsimmMOA committed:
Create WorkRecordResource.json

willsimmMOA committed:
Update WorkRecordResource.json

willsimmMOA committed:
Create OperationRecordResource.json

willsimmMOA committed:
feat: add environment data

willsimmMOA committed:
feat: environmental data

willsimmMOA committed:
Create ComponentProductType.json

willsimmMOA committed:
Create OperationProductType.json

willsimmMOA committed:
Create ProductAnalysisType.json

willsimmMOA committed:
Create productForm.json

willsimmMOA committed:
Create productType.json

willsimmMOA committed:
Create ProductIngredientType.json

willsimmMOA committed:
Create WithdrawalPeriodType.json

willsimmMOA committed:
Create RegistrationType.json

willsimmMOA committed:
Create ProductResource.json

willsimmMOA committed:
Placeholder for matterState

willsimmMOA committed:
Placeholder for quantityMeasureType

willsimmMOA committed:
feat: add placeholder for solar radiation

willsimmMOA committed:
fix: typo

willsimmMOA committed:
fix: typos

willsimmMOA committed:
Create ClimateMeasurement.json

willsimmMOA committed:
fix: path of productanalysistype

willsimmMOA committed:
fix: path of productingredienttype

willsimmMOA committed:
feat: fill out states of matter

willsimmMOA committed:
feat: build out measurement type placeholders

This will need attention, not sure how to implement this

willsimmMOA commented:
I was not sure how to build out ClimateMeasurementType and QuantityMeasurementType, at the moment they are placeholders with string for units. 

How can the correct units enum or schema be chosen within the Type schema file?

willsimmMOA committed:
feat: add resolution to measurement types

danielR9 commented on enums/uncefactVelocityUnits.json:
```diff
@@ -0,0 +1,12 @@
+{
+    "description": "A subset of units for velocity from UN/CEFACT trade facilitation recommendation 20. In order: metre per second, knot, kilometre per hour, foot per second, mile per hour (statue mile)",
```
```suggestion
    "description": "A subset of units for velocity from UN/CEFACT trade facilitation recommendation 20. In order: metre per second, knot, kilometre per hour, foot per second, mile per hour (statute mile)",
```

cookeac commented on enums/productForm.json:
```diff
@@ -0,0 +1,10 @@
+{
+    "description": "Product Form",
+    "type": "string",
+    "enum": [
```
Recommend this should become ProductPhysicalFormType.json (for physicalForm) with values:
wettable powder
water soluble powder
water dispersable granules
granules
organic solid
suspension concentrate
soluble concentrate
emulsifiable concentrate
organic suspension
vapour
gas

cookeac commented on enums/productType.json:
```diff
@@ -0,0 +1,11 @@
+    "description": "Product Type",
+    "type": "string",
+    "enum": [
+        "Chemical",
```
Recommend this should cover the types of products that could be **involved** in field activities.
For instance:
  irrigation
  seed
  plant
  fertiliser
  pesticide
  herbicide
  adjuvant
  effluent
  organic (on the theory that some organic materials are multi-factor)
  harvested product

cookeac commented on types/ClimateMeasurement.json:
```diff
@@ -0,0 +1,20 @@
+{
+    "$schema": "https://json-schema.org/draft/2020-12/schema",
```
Could we rename this file to ClimateMeasurementType.json

cookeac commented on types/ComponentProductType.json:
```diff
@@ -0,0 +1,20 @@
+            "type":  "number"
+        },
+        "percent":{
+            "type":"number"
```
Add a description: "The proportion that this product composes of the mix. Can be calculated from Wt/Vol or other measures."

```suggestion
            "type": "number",
            "description": "The proportion that this product composes of the mix. Can be calculated from Wt/Vol or other measures."
```

cookeac commented on types/OperationEnvironmentType.json:
```diff
@@ -0,0 +1,37 @@
+            "$ref": "./ClimateMeasurement.json"
+        },
+        "rainfall24hrAverage" : {
+            "$ref": "./distanceMeasure.json",
```
I know this makes sense (so you can have mm/inches) but I think it would be better documented with a description and use ClimateMeasurementType.

cookeac replied:
Decided to leave this as it is.

cookeac commented on types/OperationEnvironmentType.json:
```diff
@@ -0,0 +1,37 @@
+            "description": "Humidity"
+        },
+        "soilTemperature": {
+            "$ref": "./temperatureMeasure.json",
```
I'd rather we documented required units in description and used ClimateMeasurementType for consistency.

cookeac commented on types/OperationEnvironmentType.json:
```diff
@@ -0,0 +1,37 @@
+    "type": "object",
+    "properties": {
+        "windSpeed": {
+            "$ref": "./velocityMeasure.json",
```
I'd rather we used ClimateMeasurementType - acknowledging that the limited set of velocity units are good nevertheless - perhaps in description?

cookeac commented on types/ProductAnalysisType.json:
```diff
@@ -0,0 +1,19 @@
+{
+    "$schema": "https://json-schema.org/draft/2020-12/schema",
+    "$id": "types/ProductAnalysisType",
+    "type": "object",
```
If a product analysis is added, I suggest that **name** and **percent** should be required.

cookeac commented on types/ProductIngredientType.json:
```diff
@@ -0,0 +1,16 @@
+        "id" :{
+            "$ref" : "../types/IdentifierType.json"
+        },
+        "name": {
```
If a product ingredient is specified, **name** should be required.

cookeac commented on types/OperationEnvironmentType.json:
```diff
@@ -0,0 +1,37 @@
+            "description": "Air temperature"
+        },
+        "humidity": {
+            "$ref": "./humidityMeasure.json",
```
Better I think if we just use ClimateMeasurementType and specify expected units in the description?

cookeac commented on types/OperationEnvironmentType.json:
```diff
@@ -0,0 +1,37 @@
+            "$ref" : "../enums/windDirection.json"
+        },
+        "airTemperature": {
+            "$ref": "./temperatureMeasure.json",
```
Better I think if we just use ClimateMeasurementType and specify expected units in the description?

cookeac commented on types/OperationEnvironmentType.json:
```diff
@@ -0,0 +1,37 @@
+            "description": "Wind speed (velocity)"
+        },
+        "windDirection": {
+            "description": "Wind direction. Need to think about handling degrees.",
```
Lets have `windDirectionCompass` (the list of compass points as done here) and `windDirectionDegrees` (number).

cookeac replied:
Addressed

cookeac commented on types/OperationProductType.json:
```diff
@@ -0,0 +1,33 @@
+            "$ref":  "./QuantityMeasurementType.json"
+        },
+        "spatialMetric":{
+            "$ref":  "./QuantityMeasurementType.json"
```
Need a description on this e.g. "The spatial unit associated with application rate - e.g. (per) linear metre, square metre, etc".

cookeac replied:
```suggestion
            "$ref":  "./QuantityMeasurementType.json",
            "description": "The spatial metric to which the application rate applies e.g., [per] linear metre (MTR) or square metre (MTK)"
```

cookeac commented on types/ObservationType.json:
```diff
@@ -0,0 +1,106 @@
+        {
+            "$ref": "../types/ResourceType.json"
+        },
+        {
```
Not sure this "required" clause is quite right.
For an Observation I think we require **observationDate** and **feature**.
- featureId and holdingId are not attributes
- holding will be required in some cases, but not all.
Overall, my recommendation is to NOT have required fields in abstract types, and instead specify them in concrete types.

cookeac reviewed (commented):
I've put a number of comments in individual files.
A general naming convention used by the Geospatial project to date:
- attributes/members use camelcase with a leading lowercase (e.g., `measurementType`).
- Types use camelcase with a leading uppercase (e.g., `IdentifierType`).
- By convention we use a postfix of `Resource` on all resources that are likely to be fetched, posted or managed independently, and `Type` as the postfix on all other types that instead are embedded or used in resources (including enumerations, interestingly!).

chriiiish commented on enums/matterState.json:
```diff
@@ -0,0 +1,9 @@
+    "enum": [
+        "Solid",
+        "Liquid",
+        "Gas"
```
Should this include "Unknown" too?

chriiiish commented on enums/uncefactVelocityUnits.json:
```diff
@@ -0,0 +1,12 @@
+{
+    "description": "A subset of units for velocity from UN/CEFACT trade facilitation recommendation 20. In order: metre per second, knot, kilometre per hour, foot per second, mile per hour (statue mile)",
+    "type": "string",
+    "enum": [
+        "MTS",
+        "KNT",
+        "KMH",
+        "FR",
+        "HM",
+        "Beaufort"
```
According to [this](https://view.officeapps.live.com/op/view.aspx?src=https%3A%2F%2Funece.org%2Ffileadmin%2FDAM%2Funcefact%2Frecommendations%2Frec20%2Frec20_Rev13e_2017.xls&wdOrigin=BROWSELINK) (Line 1529) Beaufort should be `M19`. Do we want to adopt this?

![image](https://user-images.githubusercontent.com/3358790/187134599-95570c9c-063e-476f-ba70-640e23baa708.png)

```suggestion
    "description": "A subset of units for velocity from UN/CEFACT trade facilitation recommendation 20. In order: metre per second, knot, kilometre per hour, foot per second, mile per hour (statue mile), Beaufort",
    "type": "string",
    "enum": [
        "MTS",
        "KNT",
        "KMH",
        "FR",
        "HM",
        "M19"
```

cookeac replied:
```suggestion
        "M19"
```

chriiiish commented on resources/ProductResource.json:
```diff
@@ -0,0 +1,74 @@
+            "$ref": "../types/CropType.json"
+        },
+        "specificGravity": {
+            "type":"number"
```
Possible description?

```suggestion
            "type": "number",
            "description": "Relative density. The ratio of the density (mass of a unit volume) of a substance to the density of a given reference material. For liquids this is typically water at 4 degrees celcius"
```

chriiiish commented on types/ClimateMeasurement.json:
```diff
@@ -0,0 +1,20 @@
+        "value" : {
+            "type" : "number"
+        },
+        "unit" : {
+            "description" : "UNCEFACT unit",
+            "type" :"string"
```
Do we want to specify an enum for these types?

cookeac replied:
Can't until we decide what they are. At the moment its looking like maybe we just define an enumeration for each one.

chriiiish commented on types/ClimateMeasurement.json:
```diff
@@ -0,0 +1,20 @@
+        },
+        "resolution" : {
+            
+            "type" :"number"
```
Possible description (though I'm not sure if it's correct):

```suggestion
            "type" : "number",
            "description": "The number of decimal places of accuracy"
```

cookeac replied:
```suggestion
            "type" : "number",
            "description": "The smallest measurement difference (in the units) that can be discriminated given the current device settings."
```

chriiiish commented on types/ObservationType.json:
```diff
@@ -0,0 +1,106 @@
+            "description": "UTC Date (required) and time (optional) the sample was taken"
+        },
+        "phenomenonStartTime": {
+            "type": "string",
+            "description": "ISO UTC DateTime"
```
Should this have format "date-time"?
```suggestion
            "type": "string",
            "format": "date-time",
            "description": "ISO UTC DateTime"

chriiiish commented on types/ObservationType.json:
```diff
@@ -0,0 +1,106 @@
+            "description": "ISO UTC DateTime"
+        },        
+        "phenomenonEndTime": {
+            "type": "string",
+            "description": "ISO UTC DateTime"
```
Should this have format "date-time"?
```suggestion
            "type": "string",
            "format": "date-time",
            "description": "ISO UTC DateTime"

chriiiish commented on types/QuantityMeasurementType.json:
```diff
@@ -0,0 +1,21 @@
+        },
+        "unit" : {
+            "description" : "UNCEFACT unit",
+            "type" :"string"
```
Should this be an enum?

cookeac replied:
We'll need to decide what the set is, or simply specify its a code from the UN/CEFACT annex III codes. Depends where we use this.

chriiiish commented on types/RegistrationType.json:
```diff
@@ -0,0 +1,10 @@
+    "$id": "types/RegistrationType",
+    "type": "object",
+    "allOf": [
+        {"$ref": "../types/IdentifierType.json"}
```
I assume because this just inherits from `IdentifierType` and there's no extra fields that we're doing this for naming reasons?

cookeac replied:
Yes, naming reasons (documentation) only. Have put explanation in RegistrationType file now.

chriiiish commented on types/humidityMeasure.json:
```diff
@@ -0,0 +1,24 @@
+{
+    "$schema": "https://json-schema.org/draft/2020-12/schema",
+    "$id": "types/humidityMeasure",
+    "description": "Specifies a humidity measure including units (default is ).",
```
Default is....

```suggestion
    "description": "Specifies a humidity measure including units (default is relative percent).",
```

chriiiish commented on types/humidityMeasure.json:
```diff
@@ -0,0 +1,24 @@
+            "type": "string",
+            "enum": [
+                "A93",
+                "PER",
```
Line 1795 of [this](https://view.officeapps.live.com/op/view.aspx?src=https%3A%2F%2Funece.org%2Fsites%2Fdefault%2Ffiles%2F2021-06%2Frec20_Rev17e-2021.xlsx&wdOrigin=BROWSELINK) says percentage should be P1. Do we want to follow that?
![image](https://user-images.githubusercontent.com/3358790/187140574-2d68bd89-2910-4f38-81cf-40b52cc1fae0.png)

```suggestion
                "P1",
```

cookeac committed:
Update enums/uncefactVelocityUnits.json

Co-authored-by: Daniel Redmond <daniel.p.redmond@gmail.com>

cookeac committed:
Update types/ComponentProductType.json

cookeac committed:
Changes to productForm.json enumeration values

cookeac committed:
Update resources/ProductResource.json

Co-authored-by: Chris Lloyd <chriiiish@users.noreply.github.com>

cookeac committed:
Update types/ObservationType.json

Co-authored-by: Chris Lloyd <chriiiish@users.noreply.github.com>

cookeac committed:
Update description of resolution

cookeac committed:
Correct UN-CEFACT abbreviation for percent

Co-authored-by: Chris Lloyd <chriiiish@users.noreply.github.com>

cookeac committed:
Update types/humidityMeasure.json

Co-authored-by: Chris Lloyd <chriiiish@users.noreply.github.com>

cookeac committed:
Add description for spatialMetric

cookeac committed:
Set date-time format in start time

Co-authored-by: Chris Lloyd <chriiiish@users.noreply.github.com>

cookeac committed:
Correct Beaufort abbrev to M19

cookeac commented on types/OperationEnvironmentType.json:
```diff
@@ -0,0 +1,37 @@
+            "$ref": "./velocityMeasure.json",
+            "description": "Wind speed (velocity)"
+        },
+        "windDirection": {
```
```suggestion
        "windDirectionCompass": {
```

cookeac committed:
rename windDirection to windDirectionCompass

cookeac commented on enums/matterState.json:
```diff
@@ -0,0 +1,9 @@
+    "description": "Matter States",
+    "type": "string",
+    "enum": [
+        "Solid",
```
```suggestion
        "Unknown",
        "Solid",
```

cookeac committed:
Add Unknown as per suggestion

cookeac committed:
Update and rename ClimateMeasurement.json to ClimateMeasurementType.json

cookeac committed:
Update and rename productType.json to ProductTypeType.json

Added extended set of product types.

cookeac committed:
Update and rename distanceMeasure.json to DistanceMeasureType.json

cookeac committed:
Update humidityMeasure.json

cookeac committed:
Rename humidityMeasure.json to HumidityMeasureType.json

cookeac committed:
Update and rename temperatureMeasure.json to TemperatureMeasureType.json

Add resolution

cookeac committed:
Update OperationEnvironmentType.json

Change file references in $refs

cookeac committed:
Update and rename velocityMeasure.json to VelocityMeasureType.json

Renamed, changed unit to units, added resolution.

cookeac committed:
Add required fields.

cookeac committed:
Make ingredient name required

Make ingredient name required if you specify an ingredient.

cookeac committed:
Rename matterState.json to MatterStateType.json

cookeac committed:
Rename productForm.json to ProductPhysicalFormType.json

cookeac committed:
Rename windDirection.json to WindDirectionType.json

cookeac committed:
Rename workItemPriority.json to WorkItemPriorityType.json

cookeac committed:
Rename workItemStatus.json to WorkItemStatusType.json

cookeac committed:
Changed case and added "Type" to file names

Changed case and added "Type" to some $ref file names

cookeac committed:
Update $ref file names

cookeac committed:
Update OperationRecordResource.json

cookeac committed:
Update QuantityMeasurementType.json

Added description to resolution, added s to units

cookeac committed:
Added description to RegistrationType

cookeac committed:
Update ClimateMeasurementType.json

cookeac committed:
Create uncefactHumidityUnits.json

Moved humidity units into this file. Also I couldn't find H64 for specific humidity in the UN/CEFACT codes guide, but is appears to be grams of water per kilogram of dry air, so g/kg-1, or GK?

cookeac committed:
Move humidity units enum to a separate file
